### PR TITLE
add request review from specific reviewers feature in pull request

### PR DESCRIPTION
**Pull request by a1012112796**
add a way to notify specific reviewers to review like github , by add  or delet a  special type
review . The acton is  is similar to Assign ,  so many code reuse the function and items of
Assignee, but the meaning and result is different.

The Permission style is is similar to github, that only writer can add a review request from Reviewers,
but the poster can recall and remove a review request after a reviwer has revied even if he don't have
Write Premission. only manager , the poster and reviewer of a request review can remove it.

The reviewers can be requested to review contain all readers for private repo , for public, contain
all writers and watchers.

The offical Review Request will block merge if Reject can block it.

an other change: add ui otify for Assignees.

screenshots:

jta   

----
jtw

----

jtb

----

jtm

### CODE REVIEW

**Review by lafriks — CHANGES_REQUESTED** [review]

**Comment by a1012112796** [comment]
@lafriks Thanks for your checking!

File: models/review.go
Comment (guillep2k) [inline models/review.go]: The review request should replace any official reviews for that reviewer if the reviewer is "official".

File: routers/routes/routes.go
Comment (guillep2k) [inline routers/routes/routes.go]: It seems that this function doesn't allow the PR creator to request reviews unless they have special permissions.

File: web_src/js/index.js
Comment (guillep2k) [inline web_src/js/index.js]: Should this prevent default?

File: routers/repo/issue.go
Comment (a1012112796) [reply routers/repo/issue.go]: Hello, The permissions style which has been writen now is same as github, And I think it is reasonable. Because in github's seting , It's a offical manager work to Assign reviewer for a pr , which is same as Assign , Lable and Milestone.
so it's reasonable to limit to write permission. and add a re-request fun which is only for author to call after a review is exit. That's just my thinking , It should have more disscussion. or may be way can provide two mode for "owner" of arepo to choose.

File: models/review.go
Comment (a1012112796) [reply models/review.go]: sorry , I have ignored the role of this member variable before, is it used to judge whether the review is valid or not? If so, I think it should be determined by the poster of the request rather than reviewer who is requested , and when Reject can block Merge, ReviewRequest should also hve the same function. Is my think right?

File: routers/repo/issue.go
Comment (a1012112796) [reply routers/repo/issue.go]: Sorry, That's because the code stly is copy from Assign , but after thinking, I think your suggestion is necessary , because If more than two request add or remove to one reviewer in a same time , It will become to a bug, I will change them,  Thanks.

File: templates/repo/issue/view_content/sidebar.tmpl
Comment (guillep2k) [inline templates/repo/issue/view_content/sidebar.tmpl]: Is `.Assignees` correct here?

File: routers/routes/routes.go
Comment (a1012112796) [reply routers/routes/routes.go]: my thought: same as github style , a PR creator who don't have write permissions can only ask for review again after an review is posted. And the Review Request is just a official notify to reviewers, It's not meaning a special permissions like Assignees.

File: web_src/js/index.js
Comment (a1012112796) [reply web_src/js/index.js]: I'm sorry , But I don't know the meaning of your question, can you  elaborate  it, thanks

File: templates/repo/issue/view_content/sidebar.tmpl
Comment (a1012112796) [reply templates/repo/issue/view_content/sidebar.tmpl]: Because assignees  include the users who have wrie permission, that is same as I looking forward , so just reuse it but remove the poster of this PR.

File: web_src/js/index.js
Comment (lafriks) [reply web_src/js/index.js]: `event.preventDefault();`

File: routers/repo/issue.go
Comment (a1012112796) [reply routers/repo/issue.go]: Hello, I have maken my thought as an issue:#10799

File: web_src/js/index.js
Comment (a1012112796) [reply web_src/js/index.js]: thanks for your reminding, I don't know it before.@lafriks

File: models/review.go
Comment (a1012112796) [reply models/review.go]: Hello, I have modified it according to my understanding, but I am not sure whether it is correct.

File: models/review.go
Comment (a1012112796) [reply models/review.go]: sorry, I forget it , it has been deleted now.

File: models/issue_comment.go
Comment (a1012112796) [reply models/issue_comment.go:90]: I don't think it's necessary to distinguish between them because they have the same information (doer and  freviewer )， and it will make code simple if make them as one type

File: routers/routes/routes.go
Comment (guillep2k) [reply routers/routes/routes.go]: OK, I didn't know Github worked like that. It would be cool to have the option (in `app.ini`) to allow normal users to ask for reviews, because Gitea is used a lot in enterprise environments and there the requirements are a little different than in open source projects. This however could be implemented later in another PR so this one doesn't get too complicated. 👍

File: templates/repo/issue/view_content/sidebar.tmpl
Comment (guillep2k) [reply templates/repo/issue/view_content/sidebar.tmpl]: There's a difference between what users have write PR permissions and what users are able to give reviews. Who can review (and whose reviews are considered official) is calculated according to branch protection settings. If the target branch is not protected, all users with read access to the PR can review and all reviews are "official".

Users with write PR permissions are meant to perform administrative tasks on them, like closing, editing the title, etc. They may not be able to submit reviews! (except the site or repo admins, which I believe always can).

File: models/review.go
Comment (guillep2k) [inline models/review.go:489]: Interesting... this would make the request "official" if the doer is an official reviewer... I didn't see this coming, but it looks like the right thing to do. 👍

File: routers/repo/issue.go
Comment (guillep2k) [reply routers/repo/issue.go]: OK, updating my list, I still think the reviewers should be able to dismiss themselves, regardless of what permissions they have.

Also, people with write permissions should not be able to dismiss any review (I as a maintainer am not allowed to dismiss other maintainer's review here in Github). Admins/owners should.

Having said that, Gitea does have its own personality, so some tasks are not entirely resolved the same way as Github. For example, reviewer whitelists are handled separately from PR write access, so we must be careful on what we can copy and what we should reconsider.

File: templates/repo/issue/view_content/sidebar.tmpl
Comment (a1012112796) [reply templates/repo/issue/view_content/sidebar.tmpl]: I see,That mean we should allow  all reader can be requested to review except heself and the poster, But If it's a public repository , It will mean ... :cry:

Comment (a1012112796) [reply templates/repo/issue/view_content/sidebar.tmpl]: may be it shold become a search instead of a list

Comment (guillep2k) [reply templates/repo/issue/view_content/sidebar.tmpl]: Yes, I think so too.

**Comment by a1012112796** [comment]
@guillep2k hello, I have changed it follow your suggestions,  about reviewers who can be requested to review, add all readers for private repo, but olny add all watchers and writers in repo , because , readers mean all user in the platform for a public repo, but It's not necessary, Because people who don't watch it won't like to make a review for this repo.

File: models/review.go
Comment (a1012112796) [inline models/review.go]: sorry, I accidentally added a few Spaces, I wll recover it in next commit.
```suggestion
	Comments []*Comment `xorm:"-"`
```

**Comment by guillep2k** [comment]
I'm not sure using the watchers list is a good idea. A public repository's branch could have a very limited number of reviewers, and they may or may not be watching the repo. Or repositories like Gitea may have hundreds of watchers (currently 460). Maybe we could use the "watchers" function for branches that are not protected (or have no reviewer restrictions) but use the list of official/whitelisted reviewers when the target branch has one.

By the way, a public repository whose owner is a hidden organization (visible: false) should be considered like a private repository.

**Comment by a1012112796** [comment]
@guillep2k Hello, I'm sorry , but I can't fully agree with you,  first this list only show who can be requested to review not help users to choose, And in gitea any user who can read can make a review. but for the public repo, which will become a problem , Because if not have some additional limit, which will mean all users in the platform should in this list, but it's not necessary. So only add offical users (which can found in accesses table) and watchers, but I really forget the whitelisted reviewers for protect branches,  they should alse be check and add to this list. and also because this list may become very big  I add a filter box which is provided by "Semantic UI" , and may be it should become a search because load many users may cost too much time, but now I have no idea about how to do it . that's my think. thanks

**Comment by guillep2k** [comment]
As long as the whitelisted users for the target branch is included in the list, I'm OK with that.

**Comment by a1012112796** [comment]
@guillep2k Hello, after carefully check , I think It's not necessary to load  whitelisted users or groups 
because thy all can be found in accesses table, which has been load firstly. so It's not necessary to load them again. you can see these codes:

https://github.com/go-gitea/gitea/blob/209045f55b1cfdb6e52fa7a67078c1a4edb1cf4f/routers/repo/setting_protected_branch.go#L120-L125

they just use users who can found in accesses table.

 and all users in the teams of the own  organization can be found in by this way also. 

So , that means after these code. all users who can be choose into whitelisted users or the groups has been loaded by these codes:
https://github.com/a1012112796/gitea/blob/69e40d604fe77fc3e5460a9c63ee962ead7eb33a/models/repo.go#L665-L674
so it's not necessay to load again. Do you think it's right ,thanks.

File: models/review.go
Comment (guillep2k) [inline models/review.go]: For select statements we should not open a new session. We use `x` directly, unless we're in the context of another session (e.g. an UPDATE operation that spans several functions), in which case we need to pass the session as a parameter.

File: models/review.go
Comment (guillep2k) [inline models/review.go]: `Begin` is only required if we intend to `Commit` or `Rollback`. 😄

File: models/review.go
Comment (guillep2k) [inline models/review.go]: Why is the `GROUP BY` required? `max(id)` should be enough since you're not including any scalar column.

Also, you could use `x. ... .Get(&review)` instead of `x. ... .Find(...)` to get a single row. That will also check that there's only one result for you.

File: routers/repo/issue.go
Comment (guillep2k) [inline routers/repo/issue.go:385]:
```suggestion
	ctx.Data["Reviewers"], err = repo.GetReviewers(ctx.UserID, issuePosterID)
```
`ctx.UserID` always has a value, while `ctx.User` might not.

File: routers/repo/issue.go
Comment (guillep2k) [inline routers/repo/issue.go]: I think the reviewer needs access to PRs; code or issue access is not enough.

```suggestion
		pemResult = permReviewer.CanAccessAny(models.AccessModeRead, models.UnitTypePullRequests)
```

File: routers/repo/issue.go
Comment (guillep2k) [inline routers/repo/issue.go]: I think this needs to be a PR writer (not issue or code) or a whitelisted reviewer.

File: routers/repo/issue.go
Comment (guillep2k) [inline routers/repo/issue.go]: The requester shouldn't be able to revoke the request unless it's an admin. Otherwise the requester can nag the reviewers by adding/removing them repeatedly.

File: routers/repo/issue.go
Comment (guillep2k) [inline routers/repo/issue.go]: Suggestion...
```suggestion
			err = issue_service.ReviewRequest(issue, ctx.User, reviewer, event == "add")
```

File: models/review.go
Comment (a1012112796) [reply models/review.go]: Hello, I think it's not necessay to load RequestDoer full message , Because I  I just need ``doerID`` to determine user permissions in tmpl file and    `` isLegalReviewRequest``, other message are not used. I found a other item in Review type named ``OriginalAuthorID`` , but not in ``CreateReviewOptions``, can I reuse it to save ``doerID``? Thanks

File: routers/repo/issue.go
Comment (a1012112796) [reply routers/repo/issue.go:385]: ``UserID`` is not in Context type?
https://github.com/go-gitea/gitea/blob/2c25e75dca62850ddc328ed46c7b7d27b61be90b/modules/context/context.go#L32-L50

File: models/review.go
Comment (mrsdizzie) [reply models/review.go]: OriginalAuthorID is for issues/prs that were migrated from another service like Github. It stores the Github user id as OriginalAuthorID for the case that the user then creates an account on Gitea and authenticates with Github login then we can re-assign those migrated issues to their new Gitea account.

Comment (guillep2k) [reply models/review.go]: You could simply add another column to the `Comment` table.

File: routers/repo/issue.go
Comment (guillep2k) [reply routers/repo/issue.go:385]: I believe it's part of `*macaron.Context`, from which `context.Context` is derived from.

Comment (a1012112796) [reply routers/repo/issue.go:385]: ？
https://github.com/go-gitea/gitea/blob/fbb0880d3760243fd56985c5e822facd68f58ebe/vendor/gitea.com/macaron/macaron/context.go#L82-L99

Comment (guillep2k) [reply routers/repo/issue.go:385]: I'm sorry, I must be confused with something else. I swear I remember several reviews in PRs like 'change ctx.User.ID into ctx.UserID because ctx.User is not always populated'.

In any case, could you please try and see what that code does when the user is not logged in? (e.g. in an all-public repo). Maybe it can never reach there?

Comment (jolheiser) [reply routers/repo/issue.go:385]: Maybe thinking of things like `issue.RepoID` vs `issue.Repo.ID` (something like that, I might have gotten var names wrong)

Comment (a1012112796) [reply routers/repo/issue.go:385]: ok, I see, I will check them, Thanks

Comment (guillep2k) [reply routers/repo/issue.go:385]: @jolheiser Maybe it's that, mixed in with functions like `repo.MustOwner()`. Sorry about the extra work.

**Comment by guillep2k** [comment]
When possible, please try to avoid force-pushing a large PR like this. Github loses track of what changes where reviewed and review comments get orphaned. 😓

File: models/repo.go
Comment (guillep2k) [inline models/repo.go]: This command will likely fail if the number of users is large, as SQLite only supports 997 parameters. Why not use:

```
SELECT * from `user` WHERE id IN (
    SELECT user_id from access WHERE ...
    UNION
    SELECT user_id from watch WHERE ... )
```
...similar to what you did above?

Ref: https://www.w3schools.com/sql/sql_union.asp

**Comment by a1012112796** [comment]
> When possible, please try to avoid force-pushing a large PR like this. Github loses track of what changes where reviewed and review comments get orphaned. 😓

Sorry , I forgort to split new changes. do  force-pushing is to make my brabch base on the latest commit in master, and get a simple comit history. I will tI'll be careful next time.

File: routers/repo/issue.go
Comment (guillep2k) [inline routers/repo/issue.go]: We should allow official reviewers to request another review, even if they don't have AccessModeWrite. We need to call `ProtectedBranch.IsUserOfficialReviewer()` on the PR's target branch. If the requester is there, they should be allowed to request.

**Comment by guillep2k** [comment]
Sorry I kept using "whitelisted reviewers", while the proper term should have been "official reviewers". That must have been confusing. 😅

File: routers/repo/issue.go
Comment (a1012112796) [reply routers/repo/issue.go:385]: @guillep2k  Hello, I think it's not a problem to use ``user.ID`` here, because it has been protected by a permission check in issue.go on Line 819, If have't login, these code will not be used, Do you think It's right？Thanks.
https://github.com/a1012112796/gitea/blob/dd900688fd3dec1f1100cf9260d9abc7afac8a28/routers/repo/issue.go#L818-L828

File: models/repo.go
Comment (a1012112796) [reply models/repo.go]: yes , I don't know it before , I will check it, Thanks.

**Review by guillep2k — APPROVED** [review]
Wow, this was a long journey. Great job!!

**Comment by lunny** [comment]
Please resolve the conflicts.

**Comment by a1012112796** [comment]
Hello, I have  resolved the conflicts and do some ui changes, please review , Thanks.
@lunny @guillep2k @zeripath ....

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
Great work, looks good now ;)

**Comment by lafriks** [comment]
Make lgtm work

**Comment by lunny** [comment]
It's better to add the feature in comparsion doc.

**Pull request merged**
